contracts/vaults/balancer/internal/pool/BalancerComposablePoolUtils.sol

```diff
@@ -0,0 +1,451 @@
+                    sellAmount: sellAmount
+                });
+
+                deposit -= sellAmount;
+                amounts[i] = amountBought;
+                tradeIndex++;
+            }
+        }
+
+        amounts[poolContext.basePool.primaryIndex] = deposit;
+    }
+
+    function _sellToken(
+        StrategyContext memory strategyContext,
+        TradeParams memory params,
+        address sellToken,
+        address buyToken,
+        uint256 sellAmount
+    ) internal returns (uint256 buyAmount) {
+        if (DexId(params.dexId) == DexId.ZERO_EX) {
+            revert Errors.InvalidDexId(params.dexId);
+        }
+
+        ( /*uint256 amountSold */, buyAmount) = 
+            strategyContext._executeTradeExactIn({
+                params: params,
+                sellToken: sellToken,
+                buyToken: buyToken,
+                amount: sellAmount,
+                useDynamicSlippage: true
+            });
+    }
+
+    function _convertTokensToPrimary(
+        BalancerComposablePoolContext memory poolContext,
+        StrategyContext memory strategyContext,
+        ComposableRedeemParams memory params,
+        uint256[] memory exitBalances
+    ) internal returns (uint256 primaryPurchased) {
+        address[] memory tokens = poolContext.basePool.tokens;
+        uint256 primaryIndex = poolContext.basePool.primaryIndex;
+        uint256 tradeIndex;
+        for (uint256 i; i < tokens.length; i++) {
+            if (i == poolContext.bptIndex) continue;
+
+            if (i == primaryIndex) {
+                primaryPurchased += exitBalances[i];
+            } else {
+                if (exitBalances[i] > 0) {
+                    primaryPurchased += _sellToken({        
+                        strategyContext: strategyContext,
+                        params: params.redemptionTrades[tradeIndex],
+                        sellToken: tokens[i],
+                        buyToken: tokens[primaryIndex],
+                        sellAmount: exitBalances[i]
+                    });
+                }
+                tradeIndex++;
+            }
+        }
+    }
+
+    function _redeem(
+        BalancerComposablePoolContext memory poolContext,
+        StrategyContext memory strategyContext,
+        AuraStakingContext memory stakingContext,
+        uint256 strategyTokens,
+        ComposableRedeemParams memory params
+    ) internal returns (uint256 finalPrimaryBalance) {
+        uint256 bptClaim = strategyContext._redeemStrategyTokens(strategyTokens);
+        bool isSingleSidedExit = params.redemptionTrades.length == 0;
+
+        // Underlying token balances from exiting the pool
+        uint256[] memory exitBalances = _unstakeAndExitPool(
+            poolContext, stakingContext, bptClaim, params.minAmounts, isSingleSidedExit
+        );
+
+        finalPrimaryBalance = _convertTokensToPrimary(
+            poolContext, strategyContext, params, exitBalances
+        );
+    }
+
+    function _joinPoolAndStake(
```

Comment:
Hey James, I forget - are we given the option to specify a min bptMinted somewhere in the deposit flow?

Reply:
I ask because I don't see where we check to see if the prices in the balancer pool are in line. For example, when we reinvest rewards, we validate the prices in the pool before letting the transaction go forward but we don't do that here. Is the idea that we use a min bptMinted value as protection here?

Reply:
yeah, minBPT is passed in as part of the deposit params